src/async-wrap.cc

```diff
@@ -207,25 +209,23 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
   }
 
   if (ran_init_callback() && !pre_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
     pre_fn->Call(context, 0, nullptr);
     if (try_catch.HasCaught())
       FatalError("node::AsyncWrap::MakeCallback", "pre hook threw");
-    try_catch.SetVerbose(true);
   }
 
   Local<Value> ret = cb->Call(context, argc, argv);
 
-  if (try_catch.HasCaught()) {
-    return Undefined(env()->isolate());
-  }
-
   if (ran_init_callback() && !post_fn.IsEmpty()) {
-    try_catch.SetVerbose(false);
+    // TODO(trevnorris): Error reporting in async wrap post callback.
     post_fn->Call(context, 0, nullptr);
     if (try_catch.HasCaught())
       FatalError("node::AsyncWrap::MakeCallback", "post hook threw");
-    try_catch.SetVerbose(true);
+  }
+
+  // If the return value is empty then the callback threw.
+  if (ret.IsEmpty()) {
```

Comment:
If the call to `cb->Call` threw, then I would call at least the domain's `exit` function so that in the following case:

```
var domain = require('domain');

var d = domain.create();

process.on('uncaughtException', function() {
   setImmediate(function someCleanup() {});
});

d.run(function() {
  throw new Error('boom');
});
```
the function `someCleanup` doesn't run in the context of domain `d`. Currently that would happen without calling `Domain.prototype.exit` in `node::MakeCallback` because domains are exited in case of an error [only if there's a error handler](https://github.com/nodejs/node/blob/master/lib/domain.js#L71-L72). This should probably be fixed too, I'll add that to the list of GH issues to create about domains.

I'll add that to my growing list